pkg/master/controller.go

```diff
@@ -212,26 +227,10 @@ func (c *Controller) SetEndpoints(serviceName string, ip net.IP, port int) error
 		glog.Warningf("Resetting endpoints for master service %q to %v", serviceName, e)
 		return c.EndpointRegistry.UpdateEndpoints(ctx, e)
 	} else if !ipCorrect {
-		// We *always* add our own IP address; if there are too many IP
-		// addresses, we remove the ones lexicographically after our
-		// own IP address.  Given the requirements stated at the top of
-		// this function, this should cause the list of IP addresses to
-		// become eventually correct.
 		e.Subsets[0].Addresses = append(e.Subsets[0].Addresses, api.EndpointAddress{IP: ip.String()})
 		e.Subsets = endpoints.RepackSubsets(e.Subsets)
-		if addrs := &e.Subsets[0].Addresses; len(*addrs) > c.MasterCount {
```

Comment:
Right, it would be totally different and would involve e.g. locking keys in
etcd.

On Thu, Sep 24, 2015 at 10:48 AM, Clayton Coleman <notifications@github.com>
wrote:

> In pkg/master/controller.go
> <https://github.com/kubernetes/kubernetes/pull/13978#discussion_r40348998>
> :
>
> >  		e.Subsets[0].Addresses = append(e.Subsets[0].Addresses, api.EndpointAddress{IP: ip.String()})
> >  		e.Subsets = endpoints.RepackSubsets(e.Subsets)
> > -		if addrs := &e.Subsets[0].Addresses; len(*addrs) > c.MasterCount {
>
> An algorithm that didn't require master count would involve heartbeating
> the endpoints with a timestamp so that other apiservers could observe out
> of date endpoints and purge them.
> … <#1500078e94dcfd5f_>
> On Thu, Sep 24, 2015 at 1:23 PM, Daniel Smith <notifications@github.com>
> wrote: In pkg/master/controller.go <
> https://github.com/kubernetes/kubernetes/pull/13978#discussion_r40346118>
> : > e.Subsets[0].Addresses = append(e.Subsets[0].Addresses,
> api.EndpointAddress{IP: ip.String()}) > e.Subsets =
> endpoints.RepackSubsets(e.Subsets) > - if addrs := &e.Subsets[0].Addresses;
> len(*addrs) > c.MasterCount { This is critical for multiple apiservers
> coordinating to get all of their IP addresses in the endpoints list. …
> <#150006191b35e254_> On Thu, Sep 24, 2015 at 10:08 AM, jay vyas <
> notifications@github.com> wrote: In pkg/master/controller.go <
> https://github.com/kubernetes/kubernetes/pull/13978#discussion_r40344524>
> : > e.Subsets[0].Addresses = append(e.Subsets[0].Addresses,
> api.EndpointAddress{IP: ip.String()}) > e.Subsets =
> endpoints.RepackSubsets(e.Subsets) > - if addrs := &e.Subsets[0].Addresses;
> len(*addrs) > c.MasterCount { My thought is that this does nothing other
> than prune old IPs, based on if too many were already in the subset... From
> looking at the unit tests, that also seems to be the case. Maybe I've
> misread this? — Reply to this email directly or view it on GitHub <
> https://github.com/kubernetes/kubernetes/pull/13978/files#r40344524>. —
> Reply to this email directly or view it on GitHub <
> https://github.com/kubernetes/kubernetes/pull/13978/files#r40346118>.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/13978/files#r40348998>.
>